rpc/events.go

```diff
@@ -63,37 +81,373 @@ func (h *Handler) SubscribeNewHeads(ctx context.Context) (uint64, *jsonrpc.Error
+	if err != nil {
+		return nil, nil, ErrInternal
+	}
+
+	startHeader, rpcErr := h.blockHeaderByID(blockID)
+	if rpcErr != nil {
+		return nil, nil, rpcErr
+	}
+
+	if latestHeader.Number > MaxBlocksBack && startHeader.Number < latestHeader.Number-MaxBlocksBack {
+		return nil, nil, ErrTooManyBlocksBack
+	}
+
+	return startHeader, latestHeader, nil
+}
+
+// sendHistoricalHeaders sends a range of headers from the start header until the latest header
+func (h *Handler) sendHistoricalHeaders(
+	ctx context.Context,
+	startHeader *core.Header,
+	latestHeader *core.Header,
+	w jsonrpc.Conn,
+	id uint64,
+) error {
+	if startHeader == nil {
+		return nil
+	}
+
+	var err error
+
+	lastHeader := startHeader
+	for {
+		select {
+		case <-ctx.Done():
+			return ctx.Err()
+		default:
+			if err := h.sendHeader(w, lastHeader, id); err != nil {
+				return err
+			}
+
+			if lastHeader.Number == latestHeader.Number {
+				return nil
+			}
+
+			lastHeader, err = h.bcReader.BlockHeaderByNumber(lastHeader.Number + 1)
+			if err != nil {
+				return err
+			}
+		}
+	}
+}
+
+func (h *Handler) bufferNewHeaders(ctx context.Context, headerSub *feed.Subscription[*core.Header], newHeadersChan chan<- *core.Header) {
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case header := <-headerSub.Recv():
+			newHeadersChan <- header
+		}
+	}
+}
+
+func (h *Handler) processNewHeaders(ctx context.Context, newHeadersChan <-chan *core.Header, w jsonrpc.Conn, id uint64) {
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case header := <-newHeadersChan:
+			if err := h.sendHeader(w, header, id); err != nil {
+				h.log.Warnw("Error sending header", "err", err)
+				return
+			}
+		}
+	}
+}
+
+// sendHeader creates a request and sends it to the client
+func (h *Handler) sendHeader(w jsonrpc.Conn, header *core.Header, id uint64) error {
+	resp, err := json.Marshal(jsonrpc.Request{
+		Version: "2.0",
+		Method:  "starknet_subscriptionNewHeads",
+		Params: map[string]any{
+			"subscription_id": id,
+			"result":          adaptBlockHeader(header),
+		},
+	})
+	if err != nil {
+		return err
+	}
+	_, err = w.Write(resp)
+	return err
+}
+
+func (h *Handler) processReorgs(ctx context.Context, reorgSub *feed.Subscription[*sync.ReorgData], w jsonrpc.Conn, id uint64) {
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case reorg := <-reorgSub.Recv():
+			if err := h.sendReorg(w, reorg, id); err != nil {
+				h.log.Warnw("Error sending reorg", "err", err)
+				return
+			}
+		}
+	}
+}
+
+func (h *Handler) sendReorg(w jsonrpc.Conn, reorg *sync.ReorgData, id uint64) error {
+	resp, err := json.Marshal(jsonrpc.Request{
+		Version: "2.0",
+		Method:  "starknet_subscriptionReorg",
+		Params: map[string]any{
+			"subscription_id": id,
+			"result":          reorg,
+		},
+	})
+	if err != nil {
+		return err
+	}
+	_, err = w.Write(resp)
+	return err
+}
+
+func (h *Handler) SubscribeTxnStatus(ctx context.Context, txHash felt.Felt, _ *BlockID) (*SubscriptionID, *jsonrpc.Error) {
```

Comment:
As per the specs, we also need to support historical blocks. Or is it not possible?

Reply:
maybe we should return error in this case? otherwise users will be confused why it's behaving differently from other  possible impl.